doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,206 @@
+                                                                         |             +----+---------+
+                                                                         V
+                                                                       +----+---------+
+                                                                       |cBLU| obj ctx +-+
+                                                                       +----+---------+ |
+                                                                                        |
+                                                                                        V
+                                                                                       +----+---------+
+                                                                                       |cBLU| process |
+                                                                                       +----+---------+
+
+In the diagram above, two requests from the same client are being served in parallel. They share the same
+pipeline. But before one transits to the next stage, it have to wait until another leaves that stage.
+
+Better Throughput
+=================
+
+Some of us might argue, can we improve the throughput using some well known techniques?
+
+Reordering
+----------
+
+When serving requests from a client performing asynchronous I/O, OSD has the opportunity to execute
+multiple requests from the same client in parallel. But due to the second restriction, we should always
+reply the client in-order. But this does not prevent us from **executing** the requests ouf-of-order, as
+long as the reordering does not change the behavior of the execution or breaks the dependencies between
+requests.
+
+- for read requests, the "behavior" is the returned metadata and payload of object.
+- for write requests, the "behavior" is the state transition of the object storage of this PG shard and its peers.
+
+Let's consider a use case of read-read reordering, where the client sends two read requests targeting the
+same object in a row. Transposing these two read operation does not change the corresponding replies, unless
+there is a write request sneaked in after the first read request is served. But since the client sending
+read requests is not the one sending write request, the former should not be annoyed if the reordering
+changes the responses -- the object state is still consistent, and the contract of RADOS protocol is not
+broken.
+
+When it comes to read-write reordering, we cannot ensure the consistency of the behavior without checking
+if the extents of read and write operations overlap with each other. The same applies to write-write reordering.
+
+Please note, we don't ensure the ordering of execution from different clients. The RADOS application
+should use an application-level protocol to ensure this.
+
+More than 1 Visitor in a Room
```

Comment:
This isn't possible in general because it really doesn't make sense to execute multiple writes on the same object at once (we already do the only sensible version by chopping up an rbd image into multiple 4MB chunks).

For reads, it would be more like flipping the room state (once empty) to read mode, allowing as many reads in concurrently as desired, and then carefully draining those requests from the room before allowing writes again.  Some of this logic is already implemented in the obc rwlock machinery, but there'd be more work to do.

Reply:
See above for a much more in depth explanation.